### ISSUE TITLE
documentation for twisted.python.deprecate.deprecated is incomplete (and other sundries)

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3029]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3029|
|Type|defect|
|Created|2008-02-11 02:33:41Z|
|Branch|https://github.com/twisted/twisted/tree/deprecated-fixes-3029|

The `version` parameter which `deprecated` takes is not documented.

The `deprecatedVersion` attribute which `deprecated` puts on the decorated function object is not documented.  Beyond that, there is an `XXX` in the test for that attribute.

The `add_a_stack_level` nested function used in `test_deprecateEmitsWarning` is contrary to the naming convention.  A `lambda` would be as good or better than a named function here, anyway.

~~It is also a bit of a let-down that there is no trivial way to test the use of this decorator.  Instead, `assertWarns`, `DeprecationWarning`, `getDeprecationWarningString`, and some other annoying things have to be used.~~

<details><summary>Searchable metadata</summary>

```
trac-id__3029 3029
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_deprecated_fixes_3029 branches/deprecated-fixes-3029
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1202697221000000 1202697221000000
changetime__1215612974000000 1215612974000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_1"></a>|@jml commented|
|-|-|

Are you aware of `TestCase.callDeprecated`? It is a trivial way of testing the use of this decorator.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

Thanks, I didn't know about `TestCase.callDeprecated`.  I retract the last part of the ticket description.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [24193]) Branching to 'deprecated-fixes-3029'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [24194]) address problems raised in ticket descr

add docs for the version parameter and the deprecatedVersion attribute

rename add_a_stack_level


refs [#3029](https://github.com/twisted/twisted/issues/3029)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun removed owner|
|-|-|

Fixed the rest.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @exarkun|
|-|-|

Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [24245]) Merge deprecated-fixes-3029

Author: exarkun
Reviewer: therve
Fixes: [#3029](https://github.com/twisted/twisted/issues/3029)

Add documentation for the parameter to `twisted.python.deprecate.deprecated` and
for the attribute the decorator it returns adds to the function it decorates.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
